### PR TITLE
New SQL-based Azure Search index for FAC API

### PR DESCRIPTION
**Pull request by gunndabad**
Adds new Azure Search resources for a new index that gets its data from a SQL table.

Amend the `CourseSearch` to use new index and rename it to `Search` since it will be shortly be returning more than just Courses.

Fixes PTCD-353

### CODE REVIEW

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿CREATE VIEW [Pttcd].[FindACourseIndex] WITH SCHEMABINDING
+	AS SELECT
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:2]: I'd probably add a description comment that explains that this view is for the search indexer to give future readers an easier clue why this is here than the name.
```suggestion
	AS SELECT
	-- This view is depended on by the Azure Search Indexer which we use to support searching the course data

```

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		cr.DurationValue,
+		cr.StudyMode,
+		CASE WHEN cr.DeliveryMode = 1 THEN cr.AttendancePattern ELSE null END AttendancePattern,
+		cr.[National],
+		v.VenueName,
+
+		CASE WHEN ISNULL(v.AddressLine1, '') <> '' THEN v.AddressLine1 + ', ' ELSE '' END +
+			CASE WHEN ISNULL(v.AddressLine2, '') <> '' THEN v.AddressLine2 + ', ' ELSE '' END +
+			CASE WHEN ISNULL(v.Town, '') <> '' THEN v.Town + ', ' ELSE '' END +
+			CASE WHEN ISNULL(v.County, '') <> '' THEN v.County + ', ' ELSE '' END +
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:41]: https://docs.microsoft.com/en-us/sql/t-sql/language-elements/coalesce-transact-sql?view=sql-server-ver15

`COALESCE(v.County, '')` etc
shorter way of converting nulls to zero length strings

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			v.Postcode
+			AS VenueAddress,
+		v.Town AS VenueTown,
+		CASE WHEN v.VenueId IS NOT NULL THEN v.Position WHEN crr.RegionId IS NOT NULL THEN r.Position ELSE NULL END AS Position,
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:45]: Can use `coalesce` here too.

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		r.Name AS RegionName,
+
+		-- Magic numbers and logic from https://github.com/SkillsFundingAgency/dfc-providerportal-changefeedlistener/commit/608340dcfaa5c74ee8b1ae422ad902ee0c529c01#diff-5f9ef9c9ca0b0bc9af8b5c7926cfcfe31fa7e8367b9104357104ad568fbe0302R103-R104
+		CAST(CASE
+			WHEN r.RegionId IS NULL THEN 1
+			WHEN r.ParentRegionId IS NOT NULL THEN 4.5  -- Sub region
+			ELSE 2.3  -- Region
+			END AS float) AS ScoreBoost
+	FROM Pttcd.Courses c
+	JOIN Pttcd.CourseRuns cr ON c.CourseId = cr.CourseId
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:56]: I always specify join type so I don't have to remember what the default is (left/right/outer/...)

**Review by timabell — APPROVED** [review]
Could probably do with a rebase.

You might want someone who knows more about azure search to review the search bits.

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql (hunk 2 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:41]: I can't see how `COALESCE` is shorter than `ISNULL` here 😕

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql (hunk 4 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:56]: It's `inner` ;-)

Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:56]: :rocket:

File: src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql (hunk 2 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:41]: Ah I'd misunderstood what it was doing.
How about `concat_ws`? https://docs.microsoft.com/en-us/sql/t-sql/functions/concat-ws-transact-sql?view=sql-server-ver15

Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:41]: hmm, no that's only for nulls, oh well never mind. Who'd use T-SQL anyway, I wish someone would layer on a better language syntax.

Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/Views/FindACourseIndex.sql:41]: 😍 `concat_ws` is perfect here - even ignores `null`s

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql [hunk 5]
```diff
@@ -0,0 +1,160 @@
+﻿CREATE PROCEDURE [Pttcd].[RefreshFindACourseIndex]
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:1]: does dacpac deal with the alter/drop/create thing for us?

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	@CourseRunIds [Pttcd].[GuidIdTable] READONLY
+AS
+BEGIN
+
+	WITH RecordsCte AS (
+		SELECT
+			-- Id is the CourseRun's ID appended with the Region, if there is one
+			CONVERT(VARCHAR(36), cr.CourseRunId) + CASE WHEN crr.RegionId IS NOT NULL THEN '-' + crr.RegionId ELSE '' END AS Id,
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:9]: guess we can use that concat_ws thing again, just for kicks, might be a little more readable

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql (hunk 5 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:1]: Thought, now that this logic has been distilled down to a stored proc, do we want to pull some or all of this logic up to the c# layer to make it more manageable and give us compile time checks, or is there some reason this needs to stay down in the sql layer? I'm thinking that given it's one provider at a time performance might be less of an issue. Linq should be able to create an efficient query for loading all the data in one hit into memory, then we can execute a merge that's parameterized on the c# side.
I think any use of stored procs has to be strongly justified given the extra maintenance overhead the incur compared to pure c#.

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+			) AS VenueAddress,
+
+			v.Town AS VenueTown,
+			COALESCE(v.Position, r.Position) AS Position,
+			r.Name AS RegionName,
+
+			-- Magic numbers and logic from https://github.com/SkillsFundingAgency/dfc-providerportal-changefeedlistener/commit/608340dcfaa5c74ee8b1ae422ad902ee0c529c01#diff-5f9ef9c9ca0b0bc9af8b5c7926cfcfe31fa7e8367b9104357104ad568fbe0302R103-R104
+			CAST(CASE
+				WHEN r.RegionId IS NULL THEN 1
+				WHEN r.ParentRegionId IS NOT NULL THEN 4.5  -- Sub region
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:50]: You can name numbers in sql like this, a bit like a c# const.

```
declare @sub_region int = 4.5;
...
update x set y = @sub_region where z;
```
Don't worry about changing them now though, it's fine as is.

**Review by timabell — APPROVED** [review]

**Comment by timabell** [comment]
I should add, better to get this out than perfect if it's working, great work JG :rocket: :ship:

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql (hunk 5 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:1]: > does dacpac deal with the alter/drop/create thing for us?
It does

Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:1]: > Thought, now that this logic has been distilled down to a stored proc, do we want to pull some or all of this logic up to the c# layer to make it more manageable and give us compile time checks, or is there some reason this needs to stay down in the sql layer? I'm thinking that given it's one provider at a time performance might be less of an issue. Linq should be able to create an efficient query for loading all the data in one hit into memory, then we can execute a merge that's parameterized on the c# side.
I think any use of stored procs has to be strongly justified given the extra maintenance overhead the incur compared to pure c#.

We could look into that, for sure

File: src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql (hunk 7 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Database/Pttcd/StoredProcedures/RefreshFindACourseIndex.sql:50]: This whole thing needs reviewing, I think. I'm pretty sure one of the three cases here can *never* happen but it was ported over directly from the CFL for a first pass.

**Pull request merged**